ingestor-default/src/main/scala/za/co/absa/hyperdrive/ingestor/implementation/transformer/avro/confluent/AdvancedAvroToSparkConverter.scala

```diff
@@ -0,0 +1,145 @@
+            .map(metadata => StructField(f.name, schemaType.dataType, schemaType.nullable, metadata))
+            .getOrElse(StructField(f.name, schemaType.dataType, schemaType.nullable, metadataBuilderWithDefault.build()))
+        }
+
+        SchemaType(StructType(fields), nullable = false, None)
+
+      case ARRAY =>
+        val schemaType = toSqlTypeHelper(avroSchema.getElementType, existingRecordNames)
+        SchemaType(
+          ArrayType(schemaType.dataType, containsNull = schemaType.nullable),
+          nullable = false,
+          schemaType.avroType)
+
+      case MAP =>
+        val schemaType = toSqlTypeHelper(avroSchema.getValueType, existingRecordNames)
+        SchemaType(
+          MapType(StringType, schemaType.dataType, valueContainsNull = schemaType.nullable),
+          nullable = false,
+          schemaType.avroType)
+
+      case UNION =>
+        if (avroSchema.getTypes.asScala.exists(_.getType == NULL)) {
+          // In case of a union with null, eliminate it and make a recursive call
+          val remainingUnionTypes = avroSchema.getTypes.asScala.filterNot(_.getType == NULL)
+          if (remainingUnionTypes.size == 1) {
+            toSqlTypeHelper(remainingUnionTypes.head, existingRecordNames).copy(nullable = true)
+          } else {
+            toSqlTypeHelper(Schema.createUnion(remainingUnionTypes.asJava), existingRecordNames)
+              .copy(nullable = true)
+          }
+        } else avroSchema.getTypes.asScala.map(_.getType) match {
+          case Seq(t1) =>
+            toSqlTypeHelper(avroSchema.getTypes.get(0), existingRecordNames)
+          case Seq(t1, t2) if Set(t1, t2) == Set(INT, LONG) =>
+            SchemaType(LongType, nullable = false, Option(avroSchema))
+          case Seq(t1, t2) if Set(t1, t2) == Set(FLOAT, DOUBLE) =>
+            SchemaType(DoubleType, nullable = false, Option(avroSchema))
+          case _ =>
+            // Convert complex unions to struct types where field names are member0, member1, etc.
+            // This is consistent with the behavior when converting between Avro and Parquet.
+            val fields = avroSchema.getTypes.asScala.zipWithIndex.map {
+              case (s, i) =>
+                val schemaType = toSqlTypeHelper(s, existingRecordNames)
+                schemaType.avroType
+                  .map(_.toString)
+                  .map(schema => new MetadataBuilder().putString(AvroTypeKey, schema).build())
+                  .map(metadata => StructField(s"member$i", schemaType.dataType, schemaType.nullable, metadata))
+                   // All fields are nullable because only one of them is set at a time
+                  .getOrElse(StructField(s"member$i", schemaType.dataType, nullable = true))
+            }
+
+            SchemaType(StructType(fields), nullable = false, None)
+        }
+
+      case other => throw new IncompatibleSchemaException(s"Unsupported type $other")
```

Comment:
NULL type is unsupported ?

Reply:
`NULL` is handled in the `UNION` case. But I will fall back here to the default implementation as well